File: js/src/resources/text-to-sound.ts
import type { HttpClient, RequestOptions, PollingOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { CompletedTextToSoundResponse, TextToSoundParams, TextToSoundResponse, TaskCreateResponse } from '../types';

const ENDPOINT = '/api/v1/suno/text_to_sound';

/** Generates sound effects (not music) from a text description with optional looping and BPM control. */
export class TextToSound {
  constructor(private readonly http: HttpClient) {}

  /**
   * Create a text to sound task and wait until complete.
   * @param params Text to sound parameters.
   * @param options Per-request and polling overrides.
   * @returns The completed text to sound response.
   */
  async run(params: TextToSoundParams, options?: RequestOptions & PollingOptions): Promise<CompletedTextToSoundResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<TextToSoundResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedTextToSoundResponse;
  }

  /**
   * Create a text to sound task; returns immediately with a task id.
   * @param params Text to sound parameters.
   * @param options Per-request overrides.
   * @returns The task creation result.
   */
  async create(params: TextToSoundParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body: compactParams(params),
      ...options,
    });
  }

  /**
   * Fetch the current status of a text to sound task.
   * @param id The task id.
   * @param options Per-request overrides.
   * @returns The current text to sound task status.
   */
  async get(id: string, options?: RequestOptions): Promise<TextToSoundResponse> {
    return this.http.request<TextToSoundResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}

File: python/src/runapi/suno/resources/text_to_sound.py
"""Suno text_to_sound resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, RequestOptions

from .. import _validators
from ..types import CompletedTextToSoundResponse, TextToSoundResponse


class TextToSound(Resource):
    """Generate a sound effect from a text prompt."""

    ENDPOINT = "/api/v1/suno/text_to_sound"

    RESPONSE_CLASS = TextToSoundResponse
    COMPLETED_RESPONSE_CLASS = CompletedTextToSoundResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Generate sound effects and poll until it completes.

        Args:
            **params: text-to-sound parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a text-to-sound task and return immediately with an id.

        Args:
            **params: text-to-sound parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a text-to-sound task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        _validators.validate_text_to_sound(params)
